[ASTImporter] Fix infinite recurse on return type declared inside body (#68775)

danix800 opened the pull request:
Lambda without trailing auto could have return type declared inside the body too.

Fixes #68775

danix800 committed:
[ASTImporter] Fix infinite recurse on return type declared inside body (#68775)

Lambda without trailing auto could has return type declared
inside the body too.

mizvekov commented on clang/lib/AST/ASTImporter.cpp:
```diff
@@ -3649,19 +3649,15 @@ class IsTypeDeclaredInsideVisitor
 
 /// This function checks if the function has 'auto' return type that contains
 /// a reference (in any way) to a declaration inside the same function.
```
```suggestion
/// This function checks if the given function has a return type that contains
/// a reference (in any way) to a declaration inside the same function.
```

danix800 replied:
Thanks for reminding, fixed in https://github.com/llvm/llvm-project/pull/89096/commits/3f1d714a2aa4b236afe9bb1384be073fb155a2b8

mizvekov reviewed (commented):
I think it would be a good idea to double check this for performance regressions, since this case will recurse into the function every time with this patch.
Though I don't know if there is a better way to test it than to merge it and wait for it to be picked up by google.

Another idea to avoid any potential regressions, is to mark the return type as deduced anyway, with the distinction that 'auto' was not written.

danix800 committed:
Limit TypeVisitor checking to 'auto' and lambda definition only

1. Fix potential perf regression;
2. Fix comment.

danix800 commented:
> I think it would be a good idea to double check this for performance regressions, since this case will recurse into the function every time with this patch. Though I don't know if there is a better way to test it than to merge it and wait for it to be picked up by google.

For large TU it's worth considering for performance. The previous commit relaxed the `TypeVisitor` checking to all `FunctionDecl`. I reverted back to the original `auto` impl, plus an extra condition that if it's a `lambda` definition, see https://github.com/llvm/llvm-project/pull/89096/commits/3f1d714a2aa4b236afe9bb1384be073fb155a2b8

> 
> Another idea to avoid any potential regressions, is to mark the return type as deduced anyway, with the distinction that 'auto' was not written.

I tested this idea, might be like this:
```
diff --git a/clang/lib/AST/ASTImporter.cpp b/clang/lib/AST/ASTImporter.cpp
index 712958594473..988efcbef3a3 100644
--- a/clang/lib/AST/ASTImporter.cpp
+++ b/clang/lib/AST/ASTImporter.cpp
@@ -3680,24 +3680,10 @@ ExpectedDecl ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
     // E.g.: auto foo() { struct X{}; return X(); }
     // To avoid an infinite recursion when importing, create the FunctionDecl
     // with a simplified return type.
-    if (hasAutoReturnTypeDeclaredInside(D)) {
-      FromReturnTy = Importer.getFromContext().VoidTy;
-      UsedDifferentProtoType = true;
-    }
-    FunctionProtoType::ExtProtoInfo FromEPI = FromFPT->getExtProtoInfo();
-    // FunctionProtoType::ExtProtoInfo's ExceptionSpecDecl can point to the
-    // FunctionDecl that we are importing the FunctionProtoType for.
-    // To avoid an infinite recursion when importing, create the FunctionDecl
-    // with a simplified function type.
-    if (FromEPI.ExceptionSpec.SourceDecl ||
-        FromEPI.ExceptionSpec.SourceTemplate ||
-        FromEPI.ExceptionSpec.NoexceptExpr) {
-      FunctionProtoType::ExtProtoInfo DefaultEPI;
-      FromEPI = DefaultEPI;
-      UsedDifferentProtoType = true;
-    }
+    FunctionProtoType::ExtProtoInfo DefaultEPI;
+    UsedDifferentProtoType = true;
     FromTy = Importer.getFromContext().getFunctionType(
-        FromReturnTy, FromFPT->getParamTypes(), FromEPI);
+        Importer.getFromContext().VoidTy, FromFPT->getParamTypes(), DefaultEPI);
     FromTSI = Importer.getFromContext().getTrivialTypeSourceInfo(
         FromTy, D->getBeginLoc());
   }
```
but this will break two testcases from `check-clang-astmerge-exprs-cpp`:
```
Failed Tests (2):
  Clang :: ASTMerge/class-template/test.cpp
  Clang :: ASTMerge/exprs-cpp/test.cpp
```

mizvekov commented on clang/lib/AST/ASTImporter.cpp:
```diff
@@ -3647,15 +3647,19 @@ class IsTypeDeclaredInsideVisitor
   const auto *FromFPT = FromTy->getAs<FunctionProtoType>();
   assert(FromFPT && "Must be called on FunctionProtoType");
 
+  bool IsLambdaDefinition = false;
+  if (const auto *MD = dyn_cast<CXXMethodDecl>(D))
+    IsLambdaDefinition = cast<CXXRecordDecl>(MD->getDeclContext())->isLambda();
```
Thanks, this works, but if we are going this way, why not also check the lambda was defined without a return type?

danix800 replied:
> Thanks, this works, but if we are going this way, why not also check the lambda was defined without a return type?

Could you please give some direction on how to test **the lambda was defined without a return type**?

Does this mean that we need to check for possible `return` statements within lambda?

danix800 replied:
Or does this mean to test `!FromFPT->hasTrailingReturn()`?

mizvekov replied:
No, you don't need to hunt for return statements.

You can just check on the CXXMethodDecl that it's FunctionProtoType has a trailing return (`hasTrailingReturn` method).
I think that should cover it for all cases.

mizvekov replied:
Also, this problem of the return type not being wrapped on an AutoType only happens on lambdas in *C++11*.

danix800 replied:
Got it! Thanks for these useful info!

mizvekov commented:
> I tested this idea, might be like this:

I meant something else. This patch makes it so the delayed function type importation happens unconditionally.
I was aware that change would cause breakages.

What I meant is that, we have a special sugar type node that marks deduction, this AutoType, and we apply it everywhere else, except this lambda case you exposed. I was going to suggest we do the same for these lambdas. That would make things more consistent.

danix800 committed:
Check for trailing return and limit to c++11 only

The checking is refactored into a call so that if 'auto' presents, the checking
is short-circuited.

danix800 committed:
Update clang ReleaseNotes.rst

danix800 committed:
Remove debug code

mizvekov reviewed (approved):
LGTM, Thanks!

danix800 committed:
fix C++11 lang option testing
